[SYCL] disallow mutable lambdas

cperkinsintel opened the pull request:
With the latest changes in the SYCL spec kernel lambdas (and functors) must be const references. Have updated the SYCL headers and tests with this change.  Also had to make a minor adjustment to the kernel generation code in Clang and update many of the SemaSYCL tests, because their mock kernel defs needed to be updated as well. 





Signed-off-by: Chris Perkins <chris.perkins@intel.com>

AlexeySachkov commented:
> I did not update the definitions of the various reductions. Should I do so?

@vzakhari, @Pennycook: could you please answer here?

vzakhari commented:
> > I did not update the definitions of the various reductions. Should I do so?
> 
> @vzakhari, @Pennycook: could you please answer here?

@AlexeySachkov, did you want to contact @v-klochkov instead of me?

Pennycook commented:
> > I did not update the definitions of the various reductions. Should I do so?
> 
> @vzakhari, @Pennycook: could you please answer here?

I think the interfaces should be consistent.  If `parallel_for` shouldn't accept mutable lambdas, the form of `parallel_for` with support for reductions shouldn't accept mutable lambdas either.

romanovvlad commented:
@cperkinsintel  Please, click on "ready for review" button once it's ready for review.

romanovvlad commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -959,7 +961,7 @@ class __SYCL_EXPORT handler {
   detail::enable_if_t<Reduction::accessor_mode == access::mode::read_write &&
                       Reduction::has_fast_atomics && !Reduction::is_usm>
-  parallel_for(nd_range<Dims> Range, Reduction Redu, KernelType KernelFunc) {
+  parallel_for(nd_range<Dims> Range, Reduction Redu, const KernelType KernelFunc) {
```
```suggestion
  parallel_for(nd_range<Dims> Range, Reduction Redu, const KernelType &KernelFunc) {
```
The SYCL 2020 provisional says it should be a reference.

cperkinsintel replied:
Thanks!  I didn't see that.

cperkinsintel replied:
done

v-klochkov commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -720,14 +720,15 @@ class __SYCL_EXPORT handler {
   template <typename KernelName, typename KernelType>
-  __attribute__((sycl_kernel)) void kernel_single_task(KernelType KernelFunc) {
+  __attribute__((sycl_kernel)) void
+  kernel_single_task(const KernelType KernelFunc) {
```
Minor comment: kernel_single_task() is called only from single_task(), right?
If so, then 'const' in this line adds nothing because KernelType is already a const type (KernelFunc arg is 'const KernelType' type in single_task()). 
Same for kernel_parallel_for() and kernel_parallel_for_work_group().

This additional 'const' is harmless though.

cperkinsintel replied:
@v-klochkov  - what about the reductions?  They have KernelType definitions as well, and I presume the same const reference constraint will apply to them? But if they are based on these other constructs, that might be enough.  I haven't looked to know.

v-klochkov replied:
> @v-klochkov - what about the reductions? They have KernelType definitions as well, and I presume the same const reference constraint will apply to them? But if they are based on these other constructs, that might be enough. I haven't looked to know.

@Pennycook wrote the feature proposal for reduction and he already answered below that parallel_for() accepting reduction should be in sync with older parallel_for not accepting reductions. Probably, the proposal for reduction may need a minor fix to follow this changes.

v-klochkov commented:
> @v-klochkov v-klochkov requested review from elizabethandrews, Fznamznon and premanandrao as code owners now

It the result of mis-click (accidentally clicked 'Ready for review' button, which did not ask for any confirmations after the click). 
It still may be useful for code-owners to start reviewing this patch.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -730,7 +730,13 @@ getKernelInvocationKind(FunctionDecl *KernelCallerFunc) {
 
 static CXXRecordDecl *getKernelObjectType(FunctionDecl *Caller) {
-  return (*Caller->param_begin())->getType()->getAsCXXRecordDecl();
+  auto KernelParam = (*Caller->param_begin());
```
```suggestion
  auto KernelParamTy = (*Caller->param_begin())->getType();
```

cperkinsintel replied:
thanks for catching that.  Done.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -730,7 +730,13 @@ getKernelInvocationKind(FunctionDecl *KernelCallerFunc) {
-  return (*Caller->param_begin())->getType()->getAsCXXRecordDecl();
+  auto KernelParam = (*Caller->param_begin());
+  // In SYCL 2020 kernels are now passed by reference.
+  if (KernelParam->getType()->isReferenceType())
```
```suggestion
  if (KernelParamTy->isReferenceType())
```

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -730,7 +730,13 @@ getKernelInvocationKind(FunctionDecl *KernelCallerFunc) {
 }
 
 static CXXRecordDecl *getKernelObjectType(FunctionDecl *Caller) {
```
```suggestion
static const CXXRecordDecl *getKernelObjectType(FunctionDecl *Caller) {
```

erichkeane replied:
If we can, this whole function should just return a const-type.  I believe we can, it is more const correct, and it saves us a few casts.

cperkinsintel replied:
I tried that initially, but the const escapes to affect a lot of code, including some outside SemaSYCL (because of the visitor, maybe?).  This approach is the least intrusive right now.

erichkeane replied:
I actually just tried it, and other than this function, only 2 other lines are changed!  I think it is definitely worth doing here.

2 calls in SyclKernelBodyCreator need to change (Ctor and CreateKernelObjClone), plus the variable in ConstructOpenCLKernel.

I'd say it is definitely worth doing as const-casts are typically risky.

erichkeane replied:
^^ I was on an older version of the repo apparently.  I did a pull and it is a bit more involved.  If it is not non-const for a good reason, I'll likely submit a patch to do so.

cperkinsintel replied:
Thanks!  I rebased and got rid of that const cast.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -730,7 +730,13 @@ getKernelInvocationKind(FunctionDecl *KernelCallerFunc) {
+  auto KernelParam = (*Caller->param_begin());
+  // In SYCL 2020 kernels are now passed by reference.
+  if (KernelParam->getType()->isReferenceType())
+    return const_cast<CXXRecordDecl *>(
```
```suggestion
    return KernelParamTy->getPointeeCXXRecordDecl();
```

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -730,7 +730,13 @@ getKernelInvocationKind(FunctionDecl *KernelCallerFunc) {
+    return const_cast<CXXRecordDecl *>(
+        KernelParam->getType()->getPointeeCXXRecordDecl());
+  else
+    return KernelParam->getType()->getAsCXXRecordDecl(); // SYCL 1.2
```
```suggestion
    return KernelParamTy->getAsCXXRecordDecl(); // SYCL 1.2
```

cperkinsintel commented:
@erichkeane  - There is a test that ensures that the SYCL 1.2 and SYCL 2020 versions are both supported: clang/test/CodeGenSYCL/kernel-by-reference.cpp
 

Also I agree about emitting a warning. But that seems out-of-scope for this change. How does the clang front end want to handle different versions of the SYCL runtime headers? Is this the only place we'll encounter differences? I think opening a ticket for it might be better.

erichkeane commented:
> 
> 
> @erichkeane - There is a test that ensures that the SYCL 1.2 and SYCL 2020 versions are both supported: clang/test/CodeGenSYCL/kernel-by-reference.cpp
> 
> Also I agree about emitting a warning. But that seems out-of-scope for this change. How does the clang front end want to handle different versions of the SYCL runtime headers? Is this the only place we'll encounter differences? I think opening a ticket for it might be better.

I would think it is exactly in scope of this change, but I'll leave that to the code owners to make that decision.

Since we have a test that makes sure both works, I don't really see much value in changing the other 100+ tests, but I'll again leave that to the code owners to decide.

cperkinsintel commented:
@Fznamznon @elizabethandrews @premanandrao as code owners of the front-end, could one of you review that side of it?  I had to update all those ersatz test kernels to bring them up to date with the new const-ref requirement on their lambdas. That touches a lot files, but the change is trivial, just avoids deferred maintenance.  Other than that, the FE changes are very small.

premanandrao commented:
> @Fznamznon @elizabethandrews @premanandrao as code owners of the front-end, could one of you review that side of it? I had to update all those ersatz test kernels to bring them up to date with the new const-ref requirement on their lambdas. That touches a lot files, but the change is trivial, just avoids deferred maintenance. Other than that, the FE changes are very small.

I am fine with the source code changes; it is just the volume of test changes that I was uncomfortable with.
What is the long-term plan for SYCL 1.2.1 way of passing lambdas?   Will we continue to support it for a long time?  If so, I am a bitconcerned about the lack of test coverage for that style.  On the other hand, if the old style will be disallowed in the near future, then changing all the tests and having one test for completeness seems okay.

cperkinsintel commented:
@premanandrao  I don't know what the plan is for long term SYCL 1.2.1 support.   But, assuming for a second that it has to be continually supported, what would you like me to change?

The reason the volume is so high is because most of the front end tests don't actually  #include the sycl runtime headers. For good reason, because that would make the tests slow and very difficult to debug. But as a consequence of that, there are hundreds of little "fake" kernel definitions floating around in the FE tests.  In my own opinion it is better to update all of them rather than some of them.  Perhaps in the long term, a DRY solution could be found.  Should we open a ticket for that?  Or, if you'd rather I didn't update them, just let me know. I can remove all or some of them from the PR. But that seems like a deferred maintenance landmine to me.

I added a test ( clang/test/CodeGenSYCL/kernel-by-reference.cpp ) which has examples of old and new kernel to ensure the compiler is correctly compiling both. 

Just let me know what you'd like me to change.

premanandrao commented:
Good points @cperkinsintel.  I guess I don't have a right amount in mind as well.  @bader, can you share your thoughts on this?

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -730,7 +730,12 @@ getKernelInvocationKind(FunctionDecl *KernelCallerFunc) {
 
 static const CXXRecordDecl *getKernelObjectType(FunctionDecl *Caller) {
-  return (*Caller->param_begin())->getType()->getAsCXXRecordDecl();
+  auto KernelParamTy = (*Caller->param_begin())->getType();
```
Don't use auto here, according to the LLVM coding guidelines.

cperkinsintel replied:
Got it! Done.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -730,7 +730,12 @@ getKernelInvocationKind(FunctionDecl *KernelCallerFunc) {
+  // In SYCL 2020 kernels are now passed by reference.
+  if (KernelParamTy->isReferenceType())
+    return KernelParamTy->getPointeeCXXRecordDecl();
+  else
```
Don't use 'else' after a return, same guidelines.

cperkinsintel replied:
done

cperkinsintel commented:
I've rebased, updated the conflicted test, and incorporated in @erichkeane latest request. Should be good to go.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -729,7 +729,12 @@ getKernelInvocationKind(FunctionDecl *KernelCallerFunc) {
 
 static const CXXRecordDecl *getKernelObjectType(FunctionDecl *Caller) {
-  return (*Caller->param_begin())->getType()->getAsCXXRecordDecl();
+  QualType KernelParamTy = (*Caller->param_begin())->getType();
```
@elizabethandrews and @premanandrao : Did you opt to defer the diagnosing of using the wrong kernel-signature to another patch?  I don't see that comment of mine dealt with.

premanandrao replied:
@erichkeane, sorry, I mentally opted for that, but did not communicate that here.
@cperkinsintel, can you do a separate patch for that?  Pretty much every test added could potentially use the old signature, so the sooner we can get a diagnostic, the better.

cperkinsintel replied:
Sure, I'd be happy to - easily done.  What sort of thing are we emitting?   A  Warning I presume? What should it say? "Older SYCL headers encountered" ? 

`def warn_older_sycl_version : Warning<"Older SYCL headers encountered">`

erichkeane replied:
I think there are three warnings here.  

in 1.2.1 mode:
- Use of const & is a 'newer' sycl extension
-Use of pass-by-value is deprecated for 'newer' mode (akin to using noexcept and having it change type in C++17)
In Newer mode:
- Use of pass-by-value is deprecated as of SYCL-newer.

cperkinsintel replied:
How would an end-user encounter and interpret those?   We compile correctly for both old and new sycl runtime.  And the standard Clang diagnostics will warn them if they make a const violation with a functor.  

The only people writing proxy kernels is our team, for our tests. Maybe I'm misunderstanding you, but to me it seems like what we might encounter is end-user projects using SYCL 1.2 runtime headers and we should warn them of that, to encourage them to use the latest.

erichkeane replied:
The same way they interpret features based on C++ language version.  They end up being different severities of warnings in general (and of course, warnings in system headers are suppressed).

I encourage you to play around with the deprecated and extension warnings based on language features to see what I mean.

cperkinsintel replied:
I think we are talking past each other.  I'm not sure whether you are understanding me, or me not grokking you, or both.

For our users, people downloading DPCPP and writing SYCL code, there is no change to their code.  In SYCL 1.2:
`parallel_for<>([=](id){ .. do_something();});`
In SYCL 2020:
`parallel_for<>([=](id){ .. do_something();});`

But our runtime has changed such that we are now passing their lambda by reference, instead of by value.  And we support both in our code generation. So it is an opportunity for us to detect the older runtime and tell the user that it's not the latest. 

You seem to be approaching this from a different angle, but I'm unsure who the audience is.  Do we have a flag where the user gets to elect the SYCL version they want? (much like the `-std=c++17` flag).   If we did, then we would know what "mode" we are in and could advise them much like you describe. But, to my knowledge, we don't have a flag for them to elect anything like that.  Perhaps I just need a clearer example. 

Users writing functors, instead of lambdas, _will_ have to make a code change, because that operator needs to be `const`. But Clang will tell them this by virtue of the definition in sycl.hpp.

erichkeane replied:
I don't see the suggesting users to update their runtime as a valid warning here.  

However, my concern is that we expose sycl_kernel to non-runtime authors.  Therefore, we're going to get people writing sycl_kernel functions for themselves.

Therefore, we should warn if they're using the 'old' one.  If we DO have a flag based on the SYCL version (I think strict-sycl is our 1.2.1 compatibility flag), we should also warn in the other direction.

cperkinsintel replied:
I see. Hmm - as you said there is a `no-sycl-strict` flag, but it's not propagated onto the LangOpts definition, that I can see.  I'll investigate.

cperkinsintel replied:
Ok - I've updated the diagnostic to warn about deprecated pass-by-value.  Should be good. Let me know.

I looked into the "future feature" diagnostic as well, but that introduces two issues.  All the tests that I tried to bring up-to-date then have to be further updated with the expectation of that warning. That means I'll touch even more code.  But worse, when we finally switch to 2020 as the default version, all of those would have to be updated again.  (Or if not updated now, they will then). So that's creating a lot of busywork for ourselves.

erichkeane replied:
Typically you'd compile with the -Wno<warning> flag in those cases.  Just like the -Wc++17-extensions warning set.

cperkinsintel replied:
yeah, but the tests don't use that, and not sure what other consequences would be to adding it.

I think the best is to just defer the work of that "future" diagnostic until after we switch to 2020 as the default SYCL.

erichkeane replied:
As I said, I disagree.  We already HAVE switched to 2020 as the default (at least, as much as we've implemented).  As you can see, we have plenty of 2020 features by default in our compiler.

cperkinsintel replied:
My apologies if I wasn't specific enough. I'm talking about the context of these flags (`sycl-std`, `no-sycl-strict`), which I didn't even know existed until you told me.  

Presently using  `sycl-std=2020` with our clang is not supported.  However, in the update to this PR earlier today, I added that as a valid option.  So that's good.
But what I meant by "default": if no `sycl-std` is provided, then we elect  SYCL 1.2 (aka SYCL 2017) by default.  You can see that here: https://github.com/intel/llvm/blob/aaa575dd5b59c966ad769dfc9738c3d58a8d09f7/clang/lib/Driver/ToolChains/Clang.cpp#L4160

On a lark, I tried changing that to `sycl-std=2020` , but it causes a couple of unrelated  tests to fail. Probably for nothing serious. But me changing the default choice to of the Clang Front End 2020 in this PR which is ostensibly about changing the kernel lambda requirements in the runtime to be const references, seems like we are getting further and further afield.

I'm not the code owner, but to me switching the default to be 2020 and fixing the tests that fail as a result of that should be its own PR. And after that, I can come back and add the 'future' diagnostic you want, which, I believe, should be able to drop in with no further updates to kernel proxies in any tests since they've all been brought up to date.  Only the test that exists specifically to assert both kernel call conventions compile and emit the correct warnings will have to be updated. 

I'd be happy to go and open that other PR, for adding 2020 as an option, making it the default, etc.  I just assumed that decision was going to be coordinated with outside parties/developments. Let me know.

erichkeane replied:
Noone is asking you to change the default.  I simply want you to have a warning depending on which version we are in.  If the user selects 2020, but doesn't use references, thats a warning.  If the user uses 1.2.1 (or whatever our default is), and has a reference, thats also a warning.

cperkinsintel replied:
Yes, but my point is that it is easier to add the second warning after we switch the default to 2020. I'm not opposed to the warning. 
Right now, the default is 1.2.1 and most tests do not have the no-sycl-strict flag ( * ) , so if that warning is added, then a whole bunch of tests will start emitting it and need to have an expectation added, or that flag.   And when we switch the default, those tests would have to be updated again, to have that expectation removed. Alternately, I can undo the change I made to the tests bringing them up to the 2020 way of doing things. But, again, when we switch, they'll all have to be updated again.   No matter how we approach it, that second future warning causes us a lot of busywork. BUT, it's trivial to add that future warning. 

The simplest is to defer the second warning for now, and come back to it after 2020 is the default.  That's all I'm suggesting.

( * ) The clang++ driver passes the no-sycl-strict flag to the compiler. But most of these tests skip the driver and go straight to the clang_cc1  .

erichkeane replied:
Again, we use the sycl-strict flag to mean "give me warnings when I diverge from the 1.2.1 spec".  It should be a warning under sycl-strict. Which is what I said before you diverged on this.

cperkinsintel replied:
I am trying to implement them as you have asked.  My commit yesterday adds the 'deprecated' warning. Is this not correct? If not, let me know and I'll fix it. 
https://github.com/intel/llvm/pull/1785/commits/ed43c1bfc38eda9e9ddc99bd46b5f5f7c7b92507

I am not advocating for diverging from your request, only deferring the other warning that you are asking, because it's going to make a lot of busy work for someone else once the default changes to 2020. Let's defer the other warning you want until after 2020 is made the default for sycl-std.  It'll be less effort for everyone.

erichkeane replied:
Here's the problem: passing kernels by reference in SYCL 1.2.1 is INVALID and a strict compiler would emit this as an error.  Because of that, we need to warn, at least under sycl-strict.  I don't think we can defer that, since it is an ill-formed-allowed-as-extension diagnostic.

I think you're overstating the 'busy work'.  It is as simple as removing a bunch of -Wno-<warning> flags, though this even ends up not being a necessary change (since it doesn't actually break the tests to have a -Wno-<warning> flag that doesn't warn).

cperkinsintel replied:
just to be clear, when you say -Wno-  you mean the `-Wno-sycl-strict` flag, correct?

erichkeane replied:
No, you can put a warning in a warning-group.  So there would be a '-Wsycl-121-compat' group.  That is a part of the -Wsycl-strict group.

cperkinsintel replied:
Ahh - I didn't realize a warning group can be put into another.  Do we have another that is nested already, for me to see how its done?

erichkeane replied:
See DiagnosticGroups.td.  Theres a ton of examples.  Basically, the 'top level' group takes the lower groups.  "enum-conversion" would be the "sycl-strict" for example.

cperkinsintel replied:
thanks.  And that explains why I thought the flags were a big deal, and you not so much.

cperkinsintel commented:
@erichkeane  - Ok, I've added the second diagnostic, moved the emitting of it to its own function called from `ConstructOpenCLKernel`, put the warnings in their own groups, -Wno suppressed that same warning in a bunch of the tests that are now using the 2020 kernel definitions.  Let me know if there are other changes.

erichkeane commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -10969,6 +10969,12 @@ def warn_boolean_attribute_argument_is_not_valid: Warning<
     : Error<"%0 attribute cannot be applied to a "
             "static function or function in an anonymous namespace">;
+def warn_sycl_pass_by_value_deprecated 
+    : Warning<"Pass-by-value of kernel functions is deprecated in SYCL 2020.">,
```
```suggestion
    : Warning<"Passing kernel functions by value is deprecated in SYCL 2020.">,
```

cperkinsintel replied:
done

erichkeane commented on clang/include/clang/Basic/DiagnosticGroups.td:
```diff
@@ -1127,7 +1127,9 @@ def OpenMP : DiagGroup<"openmp", [
 
 // SYCL warnings
-def SyclStrict : DiagGroup<"sycl-strict">;
+def Sycl2017Conform : DiagGroup<"sycl-2017-conform">;
```
perhaps sycl-2017-compat to be more consistent with c++98-compat, et al?

cperkinsintel replied:
yes, good call.  Done.

erichkeane commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -10969,6 +10969,12 @@ def warn_boolean_attribute_argument_is_not_valid: Warning<
+    : Warning<"Pass-by-value of kernel functions is deprecated in SYCL 2020.">,
+      InGroup<Sycl2020Conform>;
+def warn_sycl_pass_by_reference_future 
+    : Warning<"Pass-by-reference of kernel functions requires SYCL 2020, and does not conform to SYCL 1.2.1">,
```
```suggestion
    : Warning<"Passing of kernel functions by reference is a SYCL 2020 extension">,
```

cperkinsintel replied:
done

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -729,7 +729,30 @@ getKernelInvocationKind(FunctionDecl *KernelCallerFunc) {
+  if (KernelParamTy->isReferenceType())
+    return KernelParamTy->getPointeeCXXRecordDecl();
+
+  // SYCL 1.2
```
```suggestion
  // SYCL 1.2.1 ?
```

cperkinsintel replied:
done

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1948,6 +1971,7 @@ void Sema::ConstructOpenCLKernel(FunctionDecl *KernelCallerFunc,
       if (LC.capturesThis() && LC.isImplicit())
         Diag(LC.getLocation(), diag::err_implicit_this_capture);
   }
+  diagKernelCallerSpecConformance(*this, KernelCallerFunc);
```
The comment I had made in the past was to move the checks from 1969 into their own function as soon as we had a 2nd thing to check.  So please make 1 function that checks both:

checkKernel<somethingorother>?  Anyone wanna bikeshed for me? @elizabethandrews  @premanandrao ?

premanandrao replied:
checkKernelParameters?

cperkinsintel replied:
checkKernelAndCaller ?

cperkinsintel replied:
checkKernelCompat ?

erichkeane replied:
> checkKernelParameters?

Its not really the Parameters, to the kernel, its the kernel object itself (plus the caller).  


> checkKernelAndCaller ?
Probably good enough for now.  Presumably someone will come up with a better idea in the future and we can rename it.

cperkinsintel replied:
done.

erichkeane commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -10969,6 +10969,12 @@ def warn_boolean_attribute_argument_is_not_valid: Warning<
     : Error<"%0 attribute cannot be applied to a "
             "static function or function in an anonymous namespace">;
+def warn_sycl_pass_by_value_deprecated 
+    : Warning<"Passing kernel functions by value is deprecated in SYCL 2020.">,
```
Nit: no periods at the end of these warnings.

cperkinsintel replied:
done

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -729,7 +729,37 @@ getKernelInvocationKind(FunctionDecl *KernelCallerFunc) {
+  }
+
+  // check that calling kernel conforms to spec
+  QualType KernelParamTy = (*Caller->param_begin())->getType();
```
Probably want an assert here that Caller->param_size() >=1.

cperkinsintel replied:
done

erichkeane commented:
@cperkinsintel : This got reverted, so you need to update this to figure out why it caused a bunch of test failures.

Fznamznon commented:
> @cperkinsintel : This got reverted, so you need to update this to figure out why it caused a bunch of test failures.

I guess he already did, see https://github.com/intel/llvm/pull/2259

erichkeane commented:
Ah, i see!  He didn't mention this PR# so I guess it didn't get picked up.
